zeppelin-interpreter/src/main/java/org/apache/zeppelin/cluster/ClusterManager.java

```diff
@@ -0,0 +1,527 @@
+import io.atomix.protocols.raft.session.CommunicationStrategy;
+import io.atomix.protocols.raft.storage.system.Configuration;
+import io.atomix.utils.net.Address;
+import io.atomix.utils.serializer.Namespace;
+import io.atomix.utils.serializer.Serializer;
+import org.apache.commons.lang.StringUtils;
+import org.apache.zeppelin.cluster.meta.ClusterMeta;
+import org.apache.zeppelin.cluster.meta.ClusterMetaEntity;
+import org.apache.zeppelin.cluster.meta.ClusterMetaOperation;
+import org.apache.zeppelin.cluster.meta.ClusterMetaType;
+import org.apache.zeppelin.cluster.protocol.LocalRaftProtocolFactory;
+import org.apache.zeppelin.cluster.protocol.RaftClientMessagingProtocol;
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.interpreter.remote.RemoteInterpreterUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.time.Instant;
+
+import java.util.Date;
+import java.util.Collections;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import static io.atomix.primitive.operation.PrimitiveOperation.operation;
+import static org.apache.zeppelin.cluster.meta.ClusterMetaOperation.DELETE_OPERATION;
+import static org.apache.zeppelin.cluster.meta.ClusterMetaOperation.PUT_OPERATION;
+import static org.apache.zeppelin.cluster.meta.ClusterMetaOperation.GET_OPERATION;
+
+/**
+ * The base class for cluster management, including the following implementations
+ * 1. RaftClient as the raft client
+ * 2. Threading to provide retry after cluster metadata submission failure
+ * 3. Cluster monitoring
+ */
+public abstract class ClusterManager {
+  private static Logger logger = LoggerFactory.getLogger(ClusterManager.class);
+
+  public final ZeppelinConfiguration zconf = ZeppelinConfiguration.create();
+
+  protected Collection<Node> clusterNodes = new ArrayList<>();
+
+  // raft
+  protected static String ZEPL_CLUSTER_ID = "ZEPL-CLUSTER";
+  protected static String ZEPL_CLIENT_ID = "ZEPL-CLIENT";
+
+  protected int raftServerPort = 0;
+
+  protected RaftClient raftClient = null;
+  protected SessionClient raftSessionClient = null;
+  protected Map<MemberId, Address> raftAddressMap = new ConcurrentHashMap<>();
+  protected LocalRaftProtocolFactory protocolFactory
+      = new LocalRaftProtocolFactory(protocolSerializer);
+  protected List<MessagingService> messagingServices = new ArrayList<>();
+  protected Collection<MemberId> clusterMemberIds = new ArrayList<MemberId>();
```

Comment:
Collection --> List (to make code style consistent)

Reply:
done